feat(run): binary-first UX — flox run <binary> [-- args] (#4177)
flox/flox: cli/flox/doc/flox-run.md

---
title: FLOX-RUN
section: 1
header: "Flox User Manuals"
...

# NAME

flox-run - run a command without installing it

# SYNOPSIS

```
flox [<general-options>] run

>      [-p=<package>]

Author: In phase 1, the package is assumed to be of the same name as the executable. You can override this with `--package` but it's optional if the names are the same.

Author: Captured in the effort's kickoff meeting notes (2026-04-16). The prototype stays end-to-end for now; the team will consider the phased delivery approach during their breakdown.

>      [-p=<package>]

Author: In future we may allow people to choose a specific version using `<name>@<version>` syntax. This can be out of scope for now.

Author: Captured in the effort under Scope Boundaries → Unknown/To Investigate as a future enhancement.

     [--reselect]
     <binary>

>      [-- <arguments>...]

Author: The `--` is only required if you want to pass options (`--opt`) to the <binary>.

If you're just passing arguments (`arg`) to the <binary>, you don't need the `--`.

We should test using `getopt` (https://linux.die.net/man/1/getopt) and make sure we're behaving in a POSIXLY_CORRECT way.

```suggestion
     [--] [<arguments> ...]
```

Author: Applied in commit 6391cf0b — synopsis changed to `[--] [<arguments>...]` and the "Passing Arguments" section now explains that `--` is only required when passing option-style arguments (e.g., `-s`, `--verbose`) to the binary. Bare arguments like URLs and filenames don't need it. TH-005 in the effort updated to reference getopt/POSIXLY_CORRECT behavior.

```

# DESCRIPTION

Run a binary from a Nix package without installing it to an environment.

`flox run` is designed for one-off invocations.
Instead of requiring the overhead of an environment, it fetches the required package and executes the binary, all in one command.

> ## Binary Lookup

Author: Let's look at https://github.com/nix-community/comma as prior art on how this can be done.

Author: Added nix-community/comma to the effort's References section.

When you run `flox run <binary>`,
Flox queries FloxHub to find which packages provide that binary.
You do not need to know the package name.
For example,
`flox run grep` will find and run the `grep` binary from
the `gnugrep` package without you needing to specify `gnugrep`.

If the binary cannot be found,
Flox will print an error with suggestions and
recommend using `--package` to specify the package directly.

## Disambiguation

If multiple packages provide the same binary
(for example, `vi` is provided by `vim`, `nvi`, and others),
Flox will prompt you to choose which package to use.
Your choice is cached so that subsequent invocations of the same
binary run silently without re-prompting.

In non-interactive contexts
(when stdin is not a terminal, such as in pipelines or CI),

> Flox will use a previously cached choice if one exists.

Reviewer: Small nit on the wording around "cache" since what's being stored is really a user's package _preference_ across invocations, I wonder if calling it a "saved preference" or "remembered choice" would be clearer. 
"Cache" implies something performance-related and ephemeral, so someone familiar might expect that and get confused.

Author: Agree about 'cache' being a confusing name, let's change that and also persist this choice in the config framework.

```suggestion
Flox will use a previous choice if one exists.
```

Author: Applied in commit 6391cf0b — renamed "cache"/"cached choice" to "saved preference" throughout the man page, error messages, and bpaf help text. Also captured TH-006 in the effort for persisting preferences in the config framework (instead of standalone TOML file).

---
title: FLOX-RUN
section: 1
header: "Flox User Manuals"
...

# NAME

flox-run - run a command without installing it

# SYNOPSIS

```
flox [<general-options>] run
     [-p=<package>]
     [--reselect]
     <binary>
     [-- <arguments>...]
```

# DESCRIPTION

Run a binary from a Nix package without installing it to an environment.

`flox run` is designed for one-off invocations where creating an
environment would be unnecessary overhead.
It creates a temporary environment behind the scenes,
installs the required package,
executes the binary,

> and cleans up when the command exits.

Reviewer 2: That feels like implementation details
```suggestion
`flox run` is designed for one-off invocations.
Instead of requiring the overhead of an environment, it fetches the required package and executes the binary, all in one command.
```

## Binary Lookup

When you run `flox run <binary>`,
Flox queries FloxHub to find which packages provide that binary.
You do not need to know the package name.
For example,
`flox run readelf` will find and run the `readelf` binary from

> the `binutils` package without you needing to specify `binutils`.

Author: ```suggestion
`flox run grep` will find and run the `grep` binary from
the `gnugrep` package without you needing to specify `gnugrep`.
```

> 

Reviewer 2: More commonly recognized example (unless maybe that would trigger prompting)
```suggestion
`flox run grep` will find and run the `grep` binary from
the `gnugrep` package without you needing to specify `gnugrep`.
```

Author: https://github.com/flox/flox/pull/4177/changes#r3089427509

If the binary cannot be found,
Flox will print an error with suggestions and
recommend using `--package` to specify the package directly.

## Disambiguation

If multiple packages provide the same binary
(for example, `vi` is provided by `vim`, `nvi`, and others),
Flox will prompt you to choose which package to use.
Your choice is cached so that subsequent invocations of the same
binary run silently without re-prompting.

In non-interactive contexts
(when stdin is not a terminal, such as in pipelines or CI),
Flox will use a previously cached choice if one exists.
If no cached choice is available,
the command will fail with a helpful error listing

> the available packages and suggesting `--package`.

Reviewer 2: It would probably be easier to just error, and particularly in non-interactive contexts users should probably be explicit

Author: Once we have the binary-to-package index, we can augment the `flox search` command to allow people to find packages that contain a specific binary (and in this error we can show the command you would run to find the relevant packages for the command you just tried to run).

Reviewer 2: - Respect config
- If there's no config and we're non-interactive, error with packages that provide the binary

Author: Applied in commit 6391cf0b — non-interactive error now shows a comma-separated list of packages inline (e.g., "Packages with this binary: vim, nvi, neovim") and uses "saved preference" naming. The no-package-found error does NOT suggest `flox search --binary` since that wouldn't help if nothing has the binary.

## Passing Arguments

Use `--` to separate `flox run` options from arguments intended
for the invoked binary.
Everything after `--` is passed directly to the binary.

# OPTIONS

## Run Options

`<binary>`
:   The name of the binary to run.
    Flox looks up which package provides this binary via FloxHub.

`-p <package>`, `--package <package>`
:   Specify the package directly,
    bypassing the binary-to-package lookup.
    This is useful when you know the package name
    or when the automatic lookup does not find the right package.
    The choice is saved to the cache

>     so that future invocations of the same binary use this package.

Reviewer 2: What binary does this run? Do I have to specify a binary separately if it's different than the package name?

Author: Currently you have to specify a `<binary>`. @nicklewis suggested having a default that's the same as the package name. I think it would be reasonable to do that.
